firmware/mcal/stm32f767/periph/analog_output.h

```diff
@@ -0,0 +1,33 @@
+/// @author Tamara Xu
+/// @date 2024-10-13
+
+#pragma once
+
+#include <cstdint>
+
+#include "shared/periph/analog_output.h"
+#include "shared/util/mappers/clamper.h"
+#include "stm32f7xx_hal.h"
+
+namespace mcal::stm32f767::periph {
+
+class AnalogOutput : public shared::periph::AnalogOutput {
+public:
+    AnalogOutput(TIM_HandleTypeDef* htim, uint32_t channel)
+        : htim_(htim), channel_(channel) {}
+
+    void SetVoltage (float voltage) override {
+        voltage_ = shared::util::Clamper<float>::Evaluate(voltage, 0, 3.3);
+        uint32_t pulse = (voltage_ * float(__HAL_TIM_GetAutoreload(htim_))) /
+                 (3.3 * 100.0f);
```

Comment:
replace `3.3` with `3.3f`

Reply:
Even though we clamp the voltage, I'm worried that a rounding error could cause the value on the right of `pulse =` to be greater than the TIM Autoreload max, and I think we would get undesired behaviour.

For example, if Autoreload is 10 bit [0-1023] and at voltage=3.3, we could have `pulse=1024`. When this value is placed in the stm register, the last 10 bits are all 0, so the duty cycle might actually be 0 instead of 100%.

Can you find a way to avoid this issue? Can we clamp `pulse` after the conversion? What would it be clamped to?